core/base/composition.cpp

```diff
@@ -33,63 +33,89 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#include <algorithm>
 
 
-namespace gko {
-namespace {
+#include <ginkgo/core/matrix/dense.hpp>
 
 
-template <typename ValueType, typename OpIterator, typename VecIterator>
-inline void allocate_vectors(OpIterator begin, OpIterator end, VecIterator res)
-{
-    for (auto it = begin; it != end; ++it) {
-        if (*res == nullptr || (*res)->get_size()[0] != (*it)->get_size()[0]) {
-            *res = matrix::Dense<ValueType>::create(
-                (*it)->get_executor(), gko::dim<2>{(*it)->get_size()[0], 1});
-        }
-        ++res;
-    }
-}
+namespace gko {
 
 
-inline const LinOp *apply_inner_operators(
+template <typename ValueType>
+std::unique_ptr<LinOp> apply_inner_operators(
     const std::vector<std::shared_ptr<const LinOp>> &operators,
-    const std::vector<std::unique_ptr<LinOp>> &intermediate, const LinOp *rhs)
+    Array<ValueType> &storage, const LinOp *rhs)
 {
-    for (auto i = operators.size() - 1; i > 0u; --i) {
-        auto solution = lend(intermediate[i - 1]);
-        operators[i]->apply(rhs, solution);
-        rhs = solution;
+    using Dense = matrix::Dense<ValueType>;
+    // determine amount of necessary storage:
+    // maximum sum of two subsequent intermediate vectors
+    // (and the out dimension of the last op if we only have one operator)
+    auto num_rhs = rhs->get_size()[1];
+    auto max_intermediate_size = std::accumulate(
+        begin(operators) + 1, end(operators) - 1,
+        operators.back()->get_size()[0],
+        [](size_type acc, std::shared_ptr<const LinOp> op) {
+            return std::max(acc, op->get_size()[0] + op->get_size()[1]);
+        });
+    auto storage_size = max_intermediate_size * num_rhs;
+    storage.resize_and_reset(storage_size);
+
+    // apply inner vectors
+    auto exec = rhs->get_executor();
+    auto data = storage.get_data();
+    // apply last operator
+    auto out_dim = gko::dim<2>{operators.back()->get_size()[0], num_rhs};
+    auto out = Dense::create(
+        exec, out_dim, Array<ValueType>::view(exec, out_dim[0] * num_rhs, data),
+        num_rhs);
+    operators.back()->apply(rhs, lend(out));
+    // apply following operators
+    // alternate intermediate vectors between beginning/end of storage
+    auto reversed_storage = true;
+    for (auto i = operators.size() - 2; i > 0; --i) {
+        // swap in and out
+        auto in = std::move(out);
+        // build new intermediate vector
+        out_dim[0] = operators[i]->get_size()[0];
+        auto out_size = out_dim[0] * num_rhs;
+        auto out_data =
+            data + (reversed_storage ? storage_size - out_size : size_type{});
+        reversed_storage = !reversed_storage;
+        out = Dense::create(exec, out_dim,
+                            Array<ValueType>::view(exec, out_size, out_data),
+                            num_rhs);
+        // apply operator
+        operators[i]->apply(lend(in), lend(out));
     }
-    return rhs;
-}
-
 
-}  // namespace
+    return std::move(out);
```

Comment:
nit: This move can hinder return value optimization. Since you are returning a `std::unique_ptr`, you can simply remove the move (you need the move in case you return a `std::shared_ptr` if `out` was a `std::unique_ptr` because there is no implicit cast for it):
```suggestion
    return out;
```

Reply:
Sorry, I have to revert that change: Intel 2018 doesn't like it.
The issue is most likely the type mismatch unique_ptr<LinOp> vs. unique_ptr<Dense>, so actually no RVO/elision can take place after all.